dlt/destinations/sql_jobs.py

```diff
@@ -486,3 +499,60 @@ def gen_merge_sql(
 
             sql.append(f"INSERT INTO {table_name}({col_str}) {select_sql};")
         return sql
+
+    @classmethod
+    def gen_scd2_sql(
+        cls, table_chain: Sequence[TTableSchema], sql_client: SqlClientBase[Any]
+    ) -> List[str]:
+        """Generates SQL statements for the `scd2` merge strategy.
+
+        The root table can be inserted into and updated.
+        Updates only take place when a record retires (because there is a new version
+        or it is deleted) and only affect the "valid to" column.
+        Child tables are insert-only.
+        """
+        sql: List[str] = []
+        root_table = table_chain[0]
+        root_table_name = sql_client.make_qualified_table_name(root_table["name"])
+        with sql_client.with_staging_dataset(staging=True):
+            staging_root_table_name = sql_client.make_qualified_table_name(root_table["name"])
+
+        # get validity column names
+        escape_id = sql_client.capabilities.escape_identifier
+        from_, to = list(map(escape_id, get_validity_column_names(root_table)))
+
+        # define values for validity columns
+        boundary_ts = current_load_package()["state"]["created_at"]
+        active_record_ts = HIGH_TS.isoformat()
+
+        # retire updated and deleted records
+        sql.append(f"""
+            UPDATE {root_table_name} SET {to} = '{boundary_ts}'
+            WHERE NOT EXISTS (
+                SELECT s._dlt_id FROM {staging_root_table_name} AS s
+                WHERE {root_table_name}._dlt_id = s._dlt_id
```

Comment:
let's add x-row-hash hint that you can attach to `_dlt_id` if it not present in the schema (same thing we do with `root_key`). why:
1. no dependency on particular name when generating sql but on hints
2. users can create their own hash during extraction and ie. just take into account 2-3 columns not all of them. also our default hash (via json dump) will be quite slow

Reply:
We could handle three user input cases:
1. `merge_key` = `None` (no merge key) ➜ assign `x-row-hash` to `_dlt_id` and generate row hash including all resource columns (the ones not generated by `dlt`)
2. `merge_key` = `col_x` (simple merge key) ➜ assign `x-row-hash` to `col_x` and use random `_dlt_id`—this can be used as "Bring Your Own Hash / Surrogate Key"
3. `merge_key` = [`col_x`, `col_y`, `col_z`] (composite merge key) ➜ assign `x-row-hash` to `_dlt_id` and generate row hash including only `col_x`, `col_y`,  and `col_z`

I suppose we would need to restrict `merge_key` changes. If we don't, we would identify unchanged records as being updated because the way their hash was calculated changed.

Reply:
Added `row_hash_column_name` configuration option in https://github.com/dlt-hub/dlt/pull/1168/commits/4124d61574d87c42230c33913d1373603af527f5. Translates to `x-row-hash` hint behind the scenes. With this, users can bring their own hash. As discussed on Slack, let's move other extensions involving `merge_key` to follow up PRs.